### PR TITLE
docs: document the merge insert operation in python

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

File: python/python/lance/dataset.py [hunk 1]
```diff
@@ -647,6 +697,58 @@ def merge_insert(
...
+
+        "Matched" records are records that exist in both the source table and
+        the target table. "Not matched" records exist only in the source table
+        (e.g. these are new data) "Not matched by source" records exist only
+        in the target table (this is old data)
+
+        The builder returned by this method can be used to customize what
+        should happen for each category of data.
+
+        Please note that there data may appear to be reordered as part of this
```
Comment (wjones127) [inline python/python/lance/dataset.py:721]:
```suggestion
        Please note that the data may appear to be reordered as part of this
```

File: python/python/lance/dataset.py [hunk 2]
```diff
@@ -647,6 +697,58 @@ def merge_insert(
...
+        operation.  This is because updated rows will be deleted from the
+        dataset and then reinserted at the end with the new values.
+
+        Parameters
+        ----------
+
+        on: Union[str, Iterable[str]]
+            A column (or columns) to join on.  This is how records from the
+            source table and target table are matched.  Typically this is some
+        kind of key or id column.
```
Comment (wjones127) [inline python/python/lance/dataset.py:731]:
```suggestion
            source table and target table are matched.  Typically this is some
            kind of key or id column.
```

File: python/python/lance/dataset.py [hunk 3]
```diff
@@ -647,6 +697,58 @@ def merge_insert(
...
+        all data where month="january")
+
+        The merge insert operation works by combining new data from a
+        **source table** with existing data in a **target table** by using a
+        join.  There are three categories of records.
+
+        "Matched" records are records that exist in both the source table and
+        the target table. "Not matched" records exist only in the source table
+        (e.g. these are new data) "Not matched by source" records exist only
+        in the target table (this is old data)
```
Comment (wjones127) [inline python/python/lance/dataset.py:716]:
```suggestion
        (e.g. these are new data). "Not matched by source" records exist only
        in the target table (this is old data).
```

**Review by wjones127 — CHANGES_REQUESTED** [review]
A few typos.

Also, do you want to add some examples to the [Writing Dataset](https://lancedb.github.io/lance/read_and_write.html#writing-lance-dataset) section? Having one code example for `insert-if-not-exists` and `upsert` each would be nice.

**Review by wjones127 — APPROVED** [review]

**Comment by westonpace** [comment]
> Also, do you want to add some examples to the [Writing Dataset](https://lancedb.github.io/lance/read_and_write.html#writing-lance-dataset) section?

I've added some examples

**Review by wjones127 — APPROVED** [review]
Excellent!

**Pull request merged**